File: ts/bullet.ts
import {
    TILE, BULLET_SIZE, CANVAS_W, CANVAS_H,
    DIR, Direction
} from './constants.js';
import { IMAGES } from './images.js';
import { Explosion } from './explosion.js';
import { playExplosionSound } from './audio.js';
import { game } from './gameState.js';
import { checkBulletMapCollision, checkBulletTankCollision } from './collision.js';
import { Tank } from './tank.js';

export class Bullet {
    x: number;
    y: number;
    dir: Direction;
    isPlayer: boolean;
    speed: number;
    width: number;
    height: number;
    alive: boolean;
    owner: Tank | null;

    constructor(x: number, y: number, dir: Direction, isPlayer: boolean) {
        this.x = x;
        this.y = y;
        this.dir = dir;
        this.isPlayer = isPlayer;
        this.speed = 4;
        this.width = BULLET_SIZE;
        this.height = BULLET_SIZE;
        this.alive = true;
        this.owner = null;
    }

    update(): void {
        if (!this.alive) return;
        if (this.dir === DIR.UP) this.y -= this.speed;
        else if (this.dir === DIR.RIGHT) this.x += this.speed;
        else if (this.dir === DIR.DOWN) this.y += this.speed;
        else if (this.dir === DIR.LEFT) this.x -= this.speed;

        if (this.x < 0 || this.y < 0 || this.x + this.width > CANVAS_W || this.y + this.height > CANVAS_H) {
            this.alive = false;
            game.explosions.push(new Explosion(this.x, this.y, 'small'));
            playExplosionSound();
            return;
        }

        if (checkBulletMapCollision(this)) return;

        checkBulletTankCollision(this);
    }

    getImage(): HTMLImageElement | null {
        const dirStr: string = ['Up', 'Right', 'Down', 'Left'][this.dir];
        return IMAGES['bullet' + dirStr] || null;
    }

    draw(ctx: CanvasRenderingContext2D): void {
        if (!this.alive) return;
        const img: HTMLImageElement | null = this.getImage();
        if (img) {
            ctx.drawImage(img, Math.floor(this.x), Math.floor(this.y), BULLET_SIZE, BULLET_SIZE);
        } else {
            ctx.fillStyle = '#fff';
            ctx.fillRect(Math.floor(this.x), Math.floor(this.y), this.width, this.height);
        }
    }
}

File: bullet.py
from constants import *
from game_state import check_bullet_map_collision, check_bullet_tank_collision
import game_state
from explosion import Explosion


class Bullet:
    def __init__(self, x, y, dir, is_player):
        self.x = x
        self.y = y
        self.dir = dir
        self.is_player = is_player
        self.speed = 4
        self.width = BULLET_SIZE
        self.height = BULLET_SIZE
        self.alive = True
        self.owner = None

    def update(self):
        if not self.alive:
            return
        if self.dir == DIR_UP:
            self.y -= self.speed
        elif self.dir == DIR_RIGHT:
            self.x += self.speed
        elif self.dir == DIR_DOWN:
            self.y += self.speed
        elif self.dir == DIR_LEFT:
            self.x -= self.speed

        if self.x < 0 or self.y < 0 or self.x + self.width > SCREEN_W or self.y + self.height > SCREEN_H:
            self.alive = False
            ex = self.x + BULLET_SIZE // 2 - TILE // 2
            ey = self.y + BULLET_SIZE // 2 - TILE // 2
            game_state.explosions.append(Explosion(ex, ey, 'small'))
            return

        if check_bullet_map_collision(self):
            return

        check_bullet_tank_collision(self)

    def get_image(self):
        dir_str = ['up', 'right', 'down', 'left'][self.dir]
        return f'bullet_{dir_str}'

    def draw(self, screen):
        if not self.alive:
            return
        img = self.get_image()
        screen.blit(img, (int(self.x), int(self.y)))
